src/solvers/smt2_incremental/object_tracking.cpp

```diff
@@ -77,6 +79,21 @@ smt_object_mapt initial_smt_object_map()
   return object_map;
 }
 
+/// This function returns true for heap allocated objects or false for stack
+/// allocated objects.
+static bool is_dynamic(const exprt &object)
+{
+  // This check corresponds to the symbols created in
+  // `goto_symext::symex_allocate`, which implements the `__CPROVER_allocate`
+  // intrinsic function used by the standard library models.
+  bool dynamic_type = object.type().get_bool(ID_C_dynamic);
+  const auto symbol = expr_try_dynamic_cast<symbol_exprt>(object);
+  bool symbol_is_dynamic =
+    symbol &&
+    has_prefix(id2string(symbol->get_identifier()), SYMEX_DYNAMIC_PREFIX);
+  return dynamic_type || symbol_is_dynamic;
```

Comment:
Yes, we could. I was thinking about readability rather than optimisation when I wrote this.

:thinking: `has_prefix`, `id2string` and `get_identifier` are all header defined functions. An optimising compiler may potentially be able to work out that these functions are all side-effect free and add in the early return...

Reply:
I've added the suggested optimisation for the moment.